### ISSUE TITLE
Allow redirecting debug logs to a file

### ISSUE THREAD

**Issue by pwmarcz** [oscarlab/graphene#1947]
I'm working with Graphene tests, and find myself switching `loader.debug_type = inline` and `loader.debug_type = none`. The default debug messages go to stdout, which makes them really hard to separate from normal program output.

I see that the default implementation opens `/dev/tty` as debug stream. There is also `loader.debug_type = file`, but it's broken: writing to a file in PAL (using `DkStreamWrite`) requires providing an absolute offset, and we pass 0, so new data overwrites the old.

## Possible solutions

* **Output logs to stderr**. Make `inline` output to `/dev/stderr` instead, or introduce `loader.debug_type = stderr`. That would already improve the situation, as it's easier to separate the logs. However, while I think it's worth doing (are there any arguments for *not* using stderr by deault? It seems to be a standard for logging), the output will be mixed with stderr from application, so it would be good to have an option to write to a file.

* **Keep track of the file offset in LibOS.** Unfortunately this requires synchronizing the offset across multiple threads, and also multiple processes. One way would be to introduce a separate thread in one of the processes, and use IPC to transfer messages to that process.

  I don't like this solution because it's complicated and pretty brittle: enabling debugging can cause things to go wrong in a harder-to-debug way; log messages might be delayed, and so on.

  If we wanted to avoid synchronizing between processes, we could write to multiple files. But I don't think that's a good solution; on the contrary, it makes it harder to debug communication between processes.

* **Allow `DkStreamWrite` to append to a file somehow.** I undestand this is not desirable to allow as it makes PAL less deterministic (although I'm not fully familiar with the argument). However, there is a `PAL_ACCESS_APPEND` option - is it supposed to do anything? There are also other resource types than "file" (e.g. device, pipe, or socket) that obviously allow such mode of output.

* **Introduce a special PAL API for logging**. In a simplest version, that would be something like `DkDebugLog(PAL_NUM count, PAL_PTR buffer)`. LibOS would just call this function, and PAL would be set up to write to stdout, stderr, or an open file (probably just by keeping a file descriptor internally and calling `write` on it).

  I like this one because I think "output a message" seems like a natural API for logging, much more so than "write to a file at an absolute offset". I think the API could also be extended with more information in the future (such as log type or level).

CC @mkow @donporter

**Comment by mkow** [comment]
I like the option with a new PAL API.
The first point looks like a separate issue, but I agree it should rather be stderr by default. Also, it would be good to be able to control the logging behavior from commandline for convenience, without changing the manifest (but this needs to be analyzed in terms of security - maybe it should be available only with some debug switch in the manifest?).

**Comment by dimakuv** [comment]
Not really related to the question of debugging, but I'm currently refactoring and simplifying the `db_devices.c` code in Linux and Linux-SGX PALs. I'm mentioning this because, as Pawel points out, the debugging functionality is implemented via a `/dev/tty` pseudof-file at LibOS level and transforms into `dev:tty`  PAL-level `device` handle (which then translates into host-level STDOUT).

I will submit the PR early next week (just refactoring). The PR would probably simplify/shed light on some of the logic that Pawel suggests.

I would actually go for option 2(**Keep track of the file offset in LibOS**), because ideally we would need to implement this anyway for correct support of append-only regular files.

On the other hand, debugging is a special feature, and I guess it must be as straight-forward and pass-through as possible (otherwise we'll end up debugging the debug-log code), so option 4 (**Introduce a special PAL API for logging**) is the most reasonable.

On additional thing to note is that this new API must be used for both LibOS and PAL logs (LibOS will use host-agnostic `DkDebugLog()` and PAL will use its host-specific `_DkDebugLog()`). Currently, LibOS and PAL use different mechanisms (even though both end up simply writing to `stdout/stderr`).

**Comment by pwmarcz** [comment]
See PR #1965.

Some further questions:

* I want to change it to stderr, maybe in the next step though.
* I agree that having the log path in manifest is pretty inconvenient. Maybe a command-line option, or env variable? No idea what would be good API here.
* There is also `_DkPrintConsole` in PAL (which prints directly to FD 2). But I guess the idea is that it prints *always*, not only when debug is enabled?
* I haven't looked at `SGX_DBG` too much yet, I guess it should also use this `DkDebugLog`? It has log levels though, should be support that in `DkDebugLog` etc.?

**Comment by mkow** [comment]
> I want to change it to stderr, maybe in the next step though.

Sounds good.
 
> I agree that having the log path in manifest is pretty inconvenient. Maybe a command-line option, or env variable? No idea what would be good API here.

I'd prefer cmdline option, but this is most likely blocked by #1053.

> There is also `_DkPrintConsole` in PAL (which prints directly to FD 2). But I guess the idea is that it prints _always_, not only when debug is enabled?

Yes, I think so.

> I haven't looked at `SGX_DBG` too much yet, I guess it should also use this `DkDebugLog`? It has log levels though, should we support that in `DkDebugLog` etc.?

No idea why there is a separate debugging utility for Linux-SGX PAL, I don't think it makes any sense. Yes, it would be nice to support loglevel + specifying the lower boundary of it in both runtime and compile time (to e.g. to allow not including debug strings in the release builds, but include warnings, but also to have debug builds which doesn't spam with info-level logs by default, but can be forced to do it using cmdline).

**Comment by pwmarcz** [comment]
So SGX doesn't really have levels, but bitmasks:

```c
#define DBG_E 0x01
#define DBG_I 0x02
#define DBG_D 0x04
#define DBG_S 0x08
#define DBG_P 0x10
#define DBG_M 0x20

#ifdef DEBUG
#define DBG_LEVEL (DBG_E | DBG_I | DBG_D | DBG_S)
#else
#define DBG_LEVEL DBG_E
#endif
```

I can guess the first three are "error", "info" and "debug", but what is `DBG_S`, `DBG_P` and `DBG_M`?

Should we replace it with more standard logging levels (error, warning, info, debug)?

**Comment by mkow** [comment]
No idea. Feel free to change everything there.

**Comment by dimakuv** [comment]
Yes, please, replace these names and remove the unneeded levels.

**Comment by dimakuv** [comment]
I think the issue was completely fixed by Pawel. @pwmarcz Feel free to reopen if there is something left.
